### PR TITLE
Versioned object support

### PR DESCRIPTION
**Pull request by shabbyrobe**
Thought I should open this as WIP in case there's any feedback, and to make the progress visible.

This will only support versioning in the memory backend for now: the iteration patterns are quite hard to fully satisfy. I think it will be tricky to support in the disk-based backends and I may opt to defer implementing this support there until a later PR if it gets too big/hard/nasty.

TODO:

- [x] Version listing in s3mem
- [x] HEAD version
- [x] Validate assumptions about prefixes and markers in listing API
- [x] Tests
- [x] List versions should work with unversioned bucket (may require further work on pagination)

### CODE REVIEW

**Comment by shabbyrobe** [comment]
This PR is getting pretty epic. I've started adding a little dump-bucket for testing assumptions about how S3 itself works. It's a bit verbose at the moment and doesn't properly clean up after itself; I'll incrementally improve it over time. For now though, it has really helped work through some of the assumptions and corner cases in the S3 API.

**Comment by shabbyrobe** [comment]
OK I'm pullin' this monster out of WIP :smile: :tada: 

There ended up being a LOT of stuff that went in to making this work. I've tried to keep the versioning stuff optional for backend implementers. Essentially, I think we can offer a bit of an a-la-carte API to GoFakeS3 for users: implement what you need (as long as you implement the Core backend) and you should be able to get up and running with something mostly S3-ish for your needs without the rest of it. I realised that by the time we have absolutely everything in here somewhere (if that ever happens!), the API surface someone who wants to make a backend would have to implement would be _insane_!

I haven't implemented versioning for the bolt or afero backends because it will take a lot of extra work to work out how, and I'd like to work on proper bucket listing pagination first (it's one of the next things on my list). This only supports versioning for the in-memory backend for now.

I kept running into situations where I was writing `FIXME` comments that said "find out what S3 actually does here", so I put together a really simple scratchpad scripting project in the `internal/s3assumer` directory. It's full of scripts that follow hopefully somewhat commented narratives to explain what detail they try to flush out of S3 itself, with a few assertions thrown in. It works against S3 itself with a real bucket; I recommend setting up a lifecycle that deletes everything it creates after a set number of days because s3assumer does not do much in the way of cleaning up after itself yet! I figured that was a bit more of a "yak shave" than I could be bothered with, but eventually I suspect I'll add it. Each test script in the assumer starts with a code number which I have then referenced in the gofakes3 code to link back to the explanation. I flushed out a _heap_ of useful detail this way, I was surprised at how much it helped.

I also had some trouble getting decent coverage reports considering the main test cases are more functional than unit tests (unit tests aren't especially useful here for the bulk of what we're doing). Go's testing tool isn't great at reporting coverage properly by default, so I've thrown together a simple script that calculates it correctly (I think) in `magefile.go`, which is run using [mage](https://github.com/magefile/mage). I noticed the `Makefile` for the first time after I added it (oops!) but I'm not sure it's easily converted from Go to shell-in-make. How would you like me to proceed? I can probably convert it to a standalone `go run` program too if you'd prefer.

From a testing point of view, I've focused primarily  on rudimentary functional coverage of all the new stuff. I think there's not nearly enough test cases around the version listing, but once again I'd like to hold off going too test-crazy until I put a proper backend-certifying tester together.

File: backend/s3mem/versionid.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+	v.mu.Unlock()
+
+	// The version IDs that come out of S3 appear to start with '3/' and follow
+	// with a base64-URL encoded blast of god knows what. There didn't appear
+	// to be any explanation of the format beyond that, but let's copy it anyway.
+	//
+	// Base64 is not sortable though, and we need our versions to be lexicographically
+	// sortable for the SkipList key, so we have to encode it as base32hex, which _is_
+	// sortable, and just pretend that it's "Base64". Phew!
```
Comment (johannesboyne) [inline backend/s3mem/versionid.go]: Hehe... ok! :D

**Comment by shabbyrobe** [comment]
> That was a big one again ;-)

Yeah, sorry 'bout that, it turned out to be a fairly large feature, much larger than I anticipated.

> Just for my understanding, is the reported codecov project coverage of 66.06% wrong or did you already changed codecov to use the magefile calculated one?

I haven't changed codecov, I didn't realise I had access to that otherwise I would've tried! 66% is kinda unsatisfyingly low, really, considering we were already at over 70. Let me see if I can add a couple more tests real quick. Unfortunately, once we get whole-project coverage, the number will drop sharply again though, as we have very patchy coverage of the other stuff in the backends and we won't recover until the backend certifying tester is up and firing (not sure when I'll get to it).

Are you happy with `mage` as a runner tool? If so, I can move things from the `Makefile` to the `magefile` now. Are there any alternatives you prefer?

File: backend/s3mem/versionid.go (hunk 1 above)
Comment (shabbyrobe) [reply backend/s3mem/versionid.go]: I know, right? It's all a bit crazy. The corner cases in the S3 API continue to astonish me!

File: backend.go [hunk 2]
```diff
@@ -69,21 +160,85 @@ type Backend interface {
...
+
+	DeleteMulti(bucketName string, objects ...string) (MultiDeleteResult, error)
+}
 
-	DeleteMulti(bucketName string, objects ...string) (DeleteResult, error)
+// VersionedBackend may be optionally implemented by a Backend in order to support
+// operations on S3 object versions.
+//
+// If you don't implement VersionedBackend, requests to GoFakeS3 that attempt to
+// make use of versions will return ErrNotImplemented.
```
[the remark this answers is not in the document]
Comment (shabbyrobe) [reply backend.go]: Ooh yeah this is a really interesting thought. I think there are definitely some operations where GoFakeS3 could make the decision and respond meaningfully if the backend doesn't implement versioning. I'll check to make sure that `GET` for the VersioningConfiguration works and just returns 'falses' for everything. `ListBucketVersions` should also work as well by simply proxying to `ListBucket`, but that's blocked waiting for `ListBucket` to get pagination support (WIP PR will be pushed in the next half an hour or so).

[the remark this answers is not in the document]
Comment (shabbyrobe) [reply backend.go]: OK GET/PUT VersioningConfiguration now work even without a versioned backend. Will have a think about the other methods. I may have to write some assumer tests to flush a few details out; I added a note into the code.

**Comment by shabbyrobe** [comment]
Wait a second, I've broken something pretty badly here. There were changes I pushed up that seem to have vanished? I force-pushed after rebasing of master last night, but I have no idea what I've done wrong as a _stack_ of fixes and extra tests are nowhere to be found.

**Comment by shabbyrobe** [comment]
Phew! Fixed it! Git fail... Ooooooops!

**Comment by shabbyrobe** [comment]
Hey wow! 64.91% is better than I expected! I think there might still be some things not being included in the report but it's definitely showing all of the backends except `s3bolt` now.

**Comment by johannesboyne** [comment]
> I got rid of the magefile because it was a pain in the backside to get it running in Circle. I've replaced it with a tools script (using the old // +build ignore trick I stole from the Go stdlib)

Now with the `// +build tools` command seems to be quite promising!

**Pull request merged**